### PR TITLE
The assembly kernel with disabling MFMA supports localsplitU for HGEMM+HPA

### PR DESCRIPTION
**Pull request by amgddm**
1. Replace bpeAB with bytesPerElem in function localSplitULocalRead
2. Replace DataType with ComputeDataType in function localSplitUReduction
3. Add assembly instruction v_add_f32 if HPA is enabled for HGEMM in function localSplitUReduction.

### CODE REVIEW

File: Tensile/KernelWriterAssembly.py [hunk 1]
```diff
@@ -9088,8 +9088,11 @@ def localSplitUReduction(self, kernel):
                          "c[%u] += c[%u], real part"%(cIdx, regIdx) )
             kStr += inst("v_add_f64", vgpr("ValuC+%u"%(cIdx+2),2), vgpr("ValuC+%u" % (regIdx+2),2), vgpr("ValuC+%u"%(cIdx+2),2), \
                          "c[%u] += c[%u], imaginary part"%(cIdx+2, regIdx+2) )
+          elif kernel["ProblemType"]["DataType"].isHalf():
```
[the remark this answers is not in the document]
Comment (nakajee) [reply Tensile/KernelWriterAssembly.py:9091]: > 

Like this (I tried 4xi8 and HHS).

  def localSplitUReduction(self, kernel):
    kStr = ""

    is_non_hpa_fp16 = kernel["ProblemType"]["ComputeDataType"].isHalf()
    elementStep = 2 if is_non_hpa_fp16 else 1
    regsPerElem = kernel["ProblemType"]["ComputeDataType"].numRegisters()

    for r in range(1, kernel["LocalSplitU"]):
      for i in range(0, kernel["NumGlobalWriteVectorsPerThread"]):
        for s in range(0, kernel["GlobalWriteVectorWidth"], elementStep):
          cIdx = int((s + i * kernel["GlobalWriteVectorWidth"]) * regsPerElem)
          regIdx = int((s + i * kernel["GlobalWriteVectorWidth"] + r * kernel["GlobalWriteVectorWidth"] * kernel["NumGlobalWriteVectorsPerThread"]) * regsPerElem)

          if kernel["ProblemType"]["ComputeDataType"].isHalf():
            kStr += inst("v_pk_add_f16", vgpr("ValuC+%u"%cIdx), vgpr("ValuC+%u" % regIdx), vgpr("ValuC+%u"%cIdx), \
                         "c[%u] += c[%u]"%(cIdx, regIdx) )
          elif kernel["ProblemType"]["ComputeDataType"].isInt32():
            kStr += inst("_v_add_i32", vgpr("ValuC+%u"%cIdx), vgpr("ValuC+%u" % regIdx), vgpr("ValuC+%u"%cIdx), \
                         "c[%u] += c[%u]"%(cIdx, regIdx))
          elif kernel["ProblemType"]["ComputeDataType"].isSingle():
            kStr += inst("v_add_f32", vgpr("ValuC+%u"%cIdx), vgpr("ValuC+%u" % regIdx), vgpr("ValuC+%u"%cIdx), \
                         "c[%u] += c[%u]"%(cIdx, regIdx))
          elif kernel["ProblemType"]["ComputeDataType"].isDouble():
            kStr += inst("v_add_f64", vgpr("ValuC+%u"%cIdx,2), vgpr("ValuC+%u" % regIdx,2), vgpr("ValuC+%u"%cIdx,2), \
                         "c[%u] += c[%u]"%(cIdx, regIdx))
          elif kernel["ProblemType"]["ComputeDataType"].isSingleComplex():
            kStr += inst("v_add_f32", vgpr("ValuC+%u"%(cIdx+0)), vgpr("ValuC+%u" % (regIdx+0)), vgpr("ValuC+%u"%(cIdx+0)), \
                         "c[%u] += c[%u], real part"%(cIdx, regIdx) )
            kStr += inst("v_add_f32", vgpr("ValuC+%u"%(cIdx+1)), vgpr("ValuC+%u" % (regIdx+1)), vgpr("ValuC+%u"%(cIdx+1)), \
                         "c[%u] += c[%u], imaginary part"%(cIdx+1, regIdx+1) )
          elif kernel["ProblemType"]["ComputeDataType"].isDoubleComplex():
            kStr += inst("v_add_f64", vgpr("ValuC+%u"%(cIdx+0),2), vgpr("ValuC+%u" % (regIdx+0),2), vgpr("ValuC+%u"%(cIdx+0),2), \
                         "c[%u] += c[%u], real part"%(cIdx, regIdx) )
            kStr += inst("v_add_f64", vgpr("ValuC+%u"%(cIdx+2),2), vgpr("ValuC+%u" % (regIdx+2),2), vgpr("ValuC+%u"%(cIdx+2),2), \
                         "c[%u] += c[%u], imaginary part"%(cIdx+2, regIdx+2) )
          else:
            # TODO: hpa_half, int8
            assert(0) # unsupported data type, need to modify here and LSU write/read code

[the remark this answers is not in the document]
Comment (nakajee) [reply Tensile/KernelWriterAssembly.py:9090]: We do not need this part
  elif kernel["ProblemType"]["DataType"].isHalf():

[the remark this answers is not in the document]
Comment (amgddm) [reply Tensile/KernelWriterAssembly.py:9090]: updated.

[the remark this answers is not in the document]
Comment (nakajee) [reply Tensile/KernelWriterAssembly.py:9090]: > updated.

I cannot see the update...

[the remark this answers is not in the document]
Comment (amgddm) [reply Tensile/KernelWriterAssembly.py]: I make a forced push again.

File: Tensile/KernelWriterAssembly.py [hunk 2]
```diff
@@ -9055,41 +9055,40 @@ def localSplitULocalRead(self, kernel):
...
         for s in range(0, kernel["GlobalWriteVectorWidth"], elementStep):
           cIdx = int((s + i * kernel["GlobalWriteVectorWidth"]) * regsPerElem)
           regIdx = int((s + i * kernel["GlobalWriteVectorWidth"] + r * kernel["GlobalWriteVectorWidth"] * kernel["NumGlobalWriteVectorsPerThread"]) * regsPerElem)
 
-          if is_non_hpa_fp16:
+          if is_fp16:
             kStr += inst("v_pk_add_f16", vgpr("ValuC+%u"%cIdx), vgpr("ValuC+%u" % regIdx), vgpr("ValuC+%u"%cIdx), \
                          "c[%u] += c[%u]"%(cIdx, regIdx) )
-          elif kernel["ProblemType"]["DataType"].isInt8x4():
+          elif kernel["ProblemType"]["ComputeDataType"].isInt8x4():
```
Comment (nakajee) [inline Tensile/KernelWriterAssembly.py:9071]: This is not correct.
ComuputeDataType should be isInt32 (as I showed in my previous comment).

**Comment by nakajee** [comment]
The code change looks OK.
Waiting for CI tests.

**Pull request merged**